oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -634,22 +663,24 @@ impl Runtime {
             let thread_ref = Arc::new(thread.clone());
 
             for reader in readers {
-                self.channels.with_channel(
-                    self.channels.get_reader_channel(*reader)?,
-                    |channel| {
-                        channel.add_waiter(thread_id, &thread_ref);
-                        Ok(())
-                    },
-                )?;
+                self.channels
+                    .with_channel(
+                        self.channels
+                            .get_reader_channel(*reader)
+                            .map_err(|e| ErroneousChannelStatus::Err(e))?,
+                        |channel| {
+                            channel.add_waiter(thread_id, &thread_ref);
+                            Ok(())
+                        },
+                    )
+                    .map_err(|e| ErroneousChannelStatus::Err(e))?;
             }
             let statuses = self.readers_statuses(node_id, &readers);
 
-            let all_unreadable = statuses.iter().all(|&s| {
-                s == ChannelReadStatus::InvalidChannel || s == ChannelReadStatus::Orphaned
-            });
+            // All statuses are either ReadReady or NotReady at this point
```

Comment:
I think the desired behaviour would be that if they do change status (e.g. one of them becomes orphaned), then the wait will be interrupted, and that information would be returned to the caller. I am not sure that's already the case though, could you check?

Reply:
You're right that the channel might become orphaned at any time; there's no synchronization to prevent that.

(However, the channel handle can't become invalid: the handle has already been validated, and so must still be valid because each Node is single-threaded &ndash; a `channel_close` in this Node's context can't happen in parallel).

So it's possible that the Node will get back a status of `NotReady` for a channel that's actually already orphaned.  But I don't think that's a problem: the Node has no way to tell the difference between that ordering of events (local-wait-starts, remote-closes, local-wait-returns-notready) and the reverse order (local-wait-starts, local-wait-returns-notready, remote-closes).  To put it another way, Node A could find that a channel has been orphaned between `wait_on_channels` and `channel_read` even if the runtime had a global lock, because Node B is on a separate thread.  (Or to put it yet another way: there is no synchronization between distinct ABI invocations, and the results of `wait_on_channels` are not guaranteed to be true even by the time it returns.)

Reply:
BTW, I *think* that the current runtime will fail to wake up waiters if a channel becomes orphaned.  Incoming #894 has a fix in 1e65d8d0a9539b6a47af3e78e6d0e37b71ad02a3